File: examples/dart/lib/billing/usecase/cancel_order.dart
// lib/billing/usecase/cancel_order.dart — Use case for cancelling orders

import '../domain/order.dart';
import '../domain/errors.dart';

class CancelOrderUseCase {
  final OrderRepository _repo;

  CancelOrderUseCase(this._repo);

  Future<void> execute(String orderId) async {
    final order = await _repo.findById(orderId);
    if (order == null) {
      throw NotFoundError('Order $orderId');
    }

    if (order.status != OrderStatus.pending) {
      throw DomainError(
        code: 'order_not_pending',
        message: 'only pending orders can be cancelled',
      );
    }

    // In a real app, we'd update the order status
  }
}

File: examples/typescript/src/billing/usecase/cancel-order.ts
// src/billing/usecase/cancel-order.ts — Use case for cancelling orders

import { OrderRepository, OrderStatus } from "../domain/order";
import { DomainError, NotFoundError } from "../domain/errors";

export class CancelOrderUseCase {
  constructor(
    private readonly orderRepo: OrderRepository,
  ) {}

  async execute(orderId: string): Promise<void> {
    const order = await this.orderRepo.findById(orderId);
    if (!order) {
      throw new NotFoundError(`order ${orderId}`);
    }

    if (order.status.value !== OrderStatus.Pending.value) {
      throw new DomainError("order_not_pending", "only pending orders can be cancelled");
    }

    order.status = OrderStatus.Cancelled;
    await this.orderRepo.save(order);
  }
}
